language/move-prover/boogie-backend/src/spec_translator.rs

```diff
@@ -801,6 +991,19 @@ impl<'env> SpecTranslator<'env> {
                 Type::TypeDomain(_) => {
                     emit!(self.writer, "{}{}: $Value", comma, var_name);
                 }
+                Type::ResourceDomain(mid, sid) => {
```

Comment:
@matbd @shazqadeer @DavidLDill Here and above we use a new range expression, `ResourceDomain`. This ranges over all resources in some memory. It has a default trigger by selecting from this memory which is generated below. 

The quantifier is as in:

```forall rsc in ResourceDomain<R>: P[rsc]```

There are some interesting type theoretical issues about this range which is why I'm not sure we can make it available to the user. Notice that even if `R` would be generic (`R<T>`) the `T` is still not part of the range, because we want to quantify over all values in memory, with arbitrary address *and* type instantiation. So the `rsc` here has a type which we currently can't represent in our type system, essentially an existential type `exists t: R<t>`. I don't fully understand yet what this means, but it seems to work well for the cases for where we use this construct for now.

Reply:
I did not follow all that.  Could you point me to some sample (for example, in the Diem framework) which uses ResourceDomain in the translation of its specification?

Reply:
Overlooked this question. As said there is no  source level support for this (yet). I'm using it to assume data invariants at verification function entry:

``
    assume  forall rsc in ResourceDomain<R>: well_formed_R(rsc)
``    

... where R in used_memory(function).

Before I did:

``
    assume  forall a: address, t1: type, ..., tn: type: well_formed_R(global<R<t1, ..., tn>>(a))
``    

This let to incomplete quantifier instantiation. My conjecture was that in difference to the v1 schema, the `ti` here boxed `$Value` instead of unboxed `$TypeValue`. So I created this new quantifier domain, which the Boogie backend can then map to quantification over unboxed types.

We have a semantic loophole indexing with types. They are build from arrays. A type instantiation is build as `empty[0 := t1][1 := t2]...`, and we use such arrays as keys to resource memory arrays. This seems to only work with syntactic equality. That is the type array `empy[0 := t]` is something else than `empty[0 := t#$TypeValue(t)]` (non-extensionality).

Given the character of our type system, I think we should be actually able to map type indexing to a fingerprint, just as is happening in the VM for data base access, and get rid of this problem this way. 

@emmazzz